### PR TITLE
[feature/lucene-knn] Add Lucene as a KNNEngine

### PR DESCRIPTION
**Pull request by jmazanec15**
### Description
Adds lucene as another k-NN engine that can be used. Adds one supported method to the engine, "hnsw", with parameters "ef_construction" and "m". In addition, adds some layers of abstraction to reduce code deduplication.

### Issues Resolved
Related #380 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/index/util/Lucene.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+     * @param methods Map of k-NN methods that the library supports
+     * @param version String representing version of library
+     */
+    Lucene(Map<String, KNNMethod> methods, String version) {
+        super(methods, version);
+    }
+
+    @Override
+    public String getExtension() {
+        throw new UnsupportedOperationException("Unable to get extension for Lucene.");
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/util/Lucene.java:61]: Can we change the message to state that  extension isn't supported for Lucene? With the current message it sounds like it's supported we couldn't read it for some reason.

File: src/main/java/org/opensearch/knn/index/util/Lucene.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @Override
+    public String getCompoundExtension() {
+        throw new UnsupportedOperationException("Unable to get compound extension for Lucene.");
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/util/Lucene.java:66]: same as for extension, can we change the message to state that  compound extension isn't supported for Lucene? With the current message it sounds like it's supported we couldn't read it for some reason.

File: src/main/java/org/opensearch/knn/index/util/Lucene.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ */
+public class Lucene extends JVMLibrary {
+
+    final static Map<String, KNNMethod> METHODS = ImmutableMap.of(
+        METHOD_HNSW,
+        KNNMethod.Builder.builder(
+            MethodComponent.Builder.builder(METHOD_HNSW)
+                .addParameter(
+                    METHOD_PARAMETER_M,
+                    new Parameter.IntegerParameter(METHOD_PARAMETER_M, KNNSettings.INDEX_KNN_DEFAULT_ALGO_PARAM_M, v -> v > 0)
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/util/Lucene.java:33]: For confirmation - is validator part of this parameter definition? I'm coming from the m and efConstruction values validation, so they fit to what Lucene is supported. Or we're passing it to Lucene the relying on validation from them?

File: src/main/java/org/opensearch/knn/index/util/KNNEngine.java [hunk 4]
```diff
@@ -54,6 +56,10 @@ public static KNNEngine getEngine(String name) {
             return FAISS;
         }
 
+        if (LUCENE.getName().equals(name)) {
+            return LUCENE;
+        }
+
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/util/KNNEngine.java:62]: nit: maybe in next PR we can refactor this to switch/case or lookup in map of <engine_name, engine instance>

File: src/main/java/org/opensearch/knn/index/util/AbstractKNNLibrary.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    protected final Map<String, KNNMethod> methods;
+    protected final String version;
+
+    /**
+     * Constructor
+     *
+     * @param methods Map of k-NN methods that the library supports
+     * @param version String representing version of library
+     */
+    AbstractKNNLibrary(Map<String, KNNMethod> methods, String version) {
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/util/AbstractKNNLibrary.java:28]: lombok AllArgsConstructor?

File: src/main/java/org/opensearch/knn/index/util/AbstractKNNLibrary.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    public String getVersion() {
+        return this.version;
+    }
+
+    @Override
+    public KNNMethod getMethod(String methodName) {
+        return this.methods.get(methodName);
+    }
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/util/AbstractKNNLibrary.java:41]: please check if both getters can be done with Lombok @Getter at class level, not sure if that works with override and if we're not overriding those in sub-classes

File: src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80CompoundFormat.java [hunk 7]
```diff
@@ -47,7 +47,10 @@ public CompoundDirectory getCompoundReader(Directory dir, SegmentInfo si, IOCont
     @Override
     public void write(Directory dir, SegmentInfo si, IOContext context) throws IOException {
         for (KNNEngine knnEngine : KNNEngine.values()) {
-            writeEngineFiles(dir, si, context, knnEngine.getExtension());
+            // Lucene is a special case. All engine writes are completely handled by it so we can skip it.
+            if (knnEngine != KNNEngine.LUCENE) {
```
[the remark this answers is not in the document]
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80CompoundFormat.java:51]: Engine is on per field basis, while codec is on whole shard basis, so I think we would still need this.

File: src/main/java/org/opensearch/knn/index/util/AbstractKNNLibrary.java (hunk 6 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/util/AbstractKNNLibrary.java:41]: Not sure how to do getter for an item in a map, but can add for version.

File: src/main/java/org/opensearch/knn/index/util/AbstractKNNLibrary.java (hunk 5 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/util/AbstractKNNLibrary.java:28]: Thought of this but wanted to keep constructor package private. Didnt realize I could add access specification. Will update.

File: src/main/java/org/opensearch/knn/index/util/KNNEngine.java (hunk 4 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/util/KNNEngine.java:62]: We could probably use switch.

File: src/main/java/org/opensearch/knn/index/util/Lucene.java (hunk 3 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/util/Lucene.java:33]: Yes, we define validation logic here for parameters. We could remove it and rely on Lucene or we could match Lucene. What do you think?

File: src/main/java/org/opensearch/knn/index/util/Lucene.java (hunk 2 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/util/Lucene.java:66]: Sure, will update.

File: src/main/java/org/opensearch/knn/index/util/Lucene.java (hunk 1 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/util/Lucene.java:61]: Sure, will update.

File: src/main/java/org/opensearch/knn/index/util/Lucene.java (hunk 3 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/knn/index/util/Lucene.java:33]: I see pros and cons of both but more inclined to have our own validation. If we declare support on our end provide documentation on what is supported and what's not it's became our burden to validate values. This came of course with the double-maintenance when we need to check on updates in Lucene and keep our code in sync, but I think such updates shouldn't happen frequently and can be done as part of upgrade to next core version.

File: src/main/java/org/opensearch/knn/index/util/AbstractKNNLibrary.java (hunk 6 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/knn/index/util/AbstractKNNLibrary.java:30]: Oh, my bad, we do additional value by key lookup. Then single getter for "version" is the way to go

**Review by martin-gaievski — APPROVED** [review]

File: src/main/java/org/opensearch/knn/index/util/Lucene.java (hunk 3 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/util/Lucene.java:33]: I think we add validation logic on our end as well. Looking at the code, the validation is the same as ours already: https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java#L76

**Review by vamshin — APPROVED** [review]
LGTM! Thanks

**Pull request merged**